ext/openssl/ossl_pkey.c

```diff
@@ -457,6 +457,43 @@ ossl_pkey_check_public_key(const EVP_PKEY *pkey)
     ossl_raise(ePKeyError, "public key missing");
 }
 
+/*
+ *  call-seq:
+ *      pkey.compare?(another_pkey) -> true|false
+ *
+ * Used primarily to check if an OpenSSL::X509::Certificate#public_key compares to its private key.
+ *
+ * == Example
+ *   x509 = OpenSSL::X509::Certificate.new(pem_encoded_certificate)
+ *   rsa_key = OpenSSL::PKey::RSA.new(pem_encoded_private_key)
+ *
+ *   rsa_key.compare?(x509.public_key) => true | false
+*/
+static VALUE
+ossl_pkey_compare(VALUE self, VALUE other)
+{
+    int ret;
+    EVP_PKEY *selfPKey;
+    EVP_PKEY *otherPKey;
+
+    GetPKey(self, selfPKey);
+    GetPKey(other, otherPKey);
+
+    // Explicitly check the key type given EVP_PKEY_ASN1_METHOD(3)
+    // docs param_cmp could return any negative number.
```

Comment:
Unfortunately, we can't use the C99 comments (`// `) until we drop support for Ruby 2.6.